Allegro hand driver

stuart-fb opened the pull request:
# Description

Adds a driver for the Allegro Hand to Polymetis.  Adapted from my earlier driver implementation for the handtelemetry repo.

Allegro Hand is built for the conda package and its tests are run in circleci.  Compiling Allegro driver no longer requires libpcanbasic, which is not currently available via conda. 

Fixes # (issue)

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [X] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [X] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [X] I want a deep design review.

## Before and After

If this PR introduces a change or fixes a bug, please add before and after screenshots (if this is causes a UX change) or before and after examples(this could be plain text, videos etc ) to demonstrate the change.

If this PR introduces a new feature, please also add examples or screenshots demonstrating the feature.

# Testing

Please describe the tests that you ran to verify your changes and how we can reproduce.

# Checklist:

- [ ] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I ran on hardware (1) all scripts in `tests/scripts`, (2) asv benchmarks.

stuart-fb commented:
Requires specifying both client and model (launch_robot.py robot_client=allegro_hand robot_model=allegro_hand) - if only the client is specified the system dies ungracefully with a gRPC error that I believe is caused by mismatched expected state vector lengths.

exhaustin commented on polymetis/polymetis/conf/robot_client/allegro_hardware.yaml:
```diff
@@ -0,0 +1,96 @@
+        - -1.66
+        - -2.80
+        - -2.97
+        - -2.80
+        - 0.08
+        - -2.80
+      joint_vel: #margin: 0.1 rad/s
+        - 2.075
+        - 2.075
+        - 2.075
+        - 2.075
+        - 2.51
+        - 2.51
+        - 2.51
+      elbow_vel: 2.075      #margin: 0.1 rad/s
+      joint_torques:  #margin: 1N for first 4 joints, 0.5N for last 3 joints
+        - 86.0
+        - 86.0
+        - 86.0
+        - 86.0
+        - 11.5
+        - 11.5
+        - 11.5
+
+    collision_behavior:
+      lower_torque: [20.0, 20.0, 20.0, 20.0, 20.0, 20.0, 20.0]
+      upper_torque: [20.0, 20.0, 20.0, 20.0, 20.0, 20.0, 20.0]
+      lower_force: [20.0, 20.0, 20.0, 20.0, 20.0, 20.0]
+      upper_force: [20.0, 20.0, 20.0, 20.0, 20.0, 20.0]
+
+    safety_controller:
+      is_active: true
+      margins: # margin from hard safety limits at which safety controllers start to kick in
+        cartesian_pos: 0.05
+        joint_pos: 0.2
+        joint_vel: 0.5
+      stiffness:
+        cartesian_pos: 200.0
+        joint_pos: 50.0
+        joint_vel: 20.0
```
CMIIW, but since these fields aren't being used within the allegro client, there shouldn't be a need to include them in the config specific to allegro, right?

stuart-fb replied:
Right - I'll delete the unused ones.

exhaustin commented on polymetis/polymetis/python/polymetis/robot_client/executable_robot_client.py:
```diff
@@ -66,6 +66,8 @@ def run(self):
 
                 # Find path to executable
                 path_to_exec = which(self.executable_cfg.exec)
+                log.warn("Executing " + path_to_exec)
```
```suggestion
                log.info("Executing " + path_to_exec)
```

stuart-fb replied:
Oh, this was just a debug line - I'll drop it entirely, thanks for catching it.

exhaustin commented on polymetis/polymetis/src/clients/allegro_hand_client/allegro_hand_client.cpp:
```diff
@@ -0,0 +1,213 @@
+  spdlog::error("Max recovery attempts exhausted.");
+}
+
+void AllegroHandTorqueControlClient::updateServerCommand(
+    /*
+     * Send robot states and receive torque command via a request to the
+     * controller server.
+     */
+    const AllegroHandState &robot_state,
+    std::array<double, NUM_DOFS> &torque_out) {
+  // Record robot states
+  for (int i = 0; i < NUM_DOFS; i++) {
+    robot_state_.set_joint_positions(i, robot_state.q[i]);
+    robot_state_.set_joint_velocities(i, robot_state.dq[i]);
+  }
+  setTimestampToNow(robot_state_.mutable_timestamp());
+
+  // Retrieve torques
+  grpc::ClientContext context;
+  long int pre_update_ns = getNanoseconds();
+  status_ = stub_->ControlUpdate(&context, robot_state_, &torque_command_);
+  long int post_update_ns = getNanoseconds();
+  if (!status_.ok()) {
+    std::string error_msg = "ControlUpdate rpc failed. ";
+    throw std::runtime_error(error_msg + status_.error_message());
+  }
+
+  robot_state_.set_prev_controller_latency_ms(
+      float(post_update_ns - pre_update_ns) / 1e6);
+
+  assert(torque_command_.joint_torques_size() == NUM_DOFS);
+  for (int i = 0; i < NUM_DOFS; i++) {
+    torque_out[i] = torque_command_.joint_torques(i);
+    robot_state_.set_prev_joint_torques_computed(
+        i, torque_command_.joint_torques(i));
+  }
+}
```
A lot of code in here and in other places of the same file is copied from the `FrankaTorqueControlClient`. It should be ok for now to get the Allegro hand working with Polymetis asap, but we should make an issue or a task in the icebox to deduplicate by maybe having a `TorqueControlClient` base class.

stuart-fb replied:
Yup!

exhaustin replied:
Created an issue: #1039

exhaustin reviewed (changes requested):
I don't have a lot of context regarding the Allegro hand, but in general the interface with the Polymetis server looks good and the unit tests look reasonable. I left some minor comments that should be easily addressed.

I would include an integration test using the mocked Allegro client. If you don't have bandwidth, feel free to delegate that to me.

stuart-fb commented on polymetis/polymetis/conf/robot_client/allegro_hardware.yaml:
```diff
@@ -0,0 +1,96 @@
+defaults:
+  - robot_client/executable_cfg/velocity_filter: filters/butterworth_filters
```
fake_clock became a submodule.

stuart-fb commented:
I updated to using SocketCAN, which allows me to build and test the Allegro Client in CircleCI and added some docs about setting up SocketCAN interfaces on a machine.

> I would include an integration test using the mocked Allegro client. If you don't have bandwidth, feel free to delegate that to me.
  @exhaustin if you have bandwidth to add an integration test, I'd appreciate it.

Unless there are additional changes requests, I think this is ready to go.

exhaustin reviewed (approved):
Added an integration test to ensure the server launches with a mocked hw client, and is able to take commands from the user.
Looks good to merge!
